mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/SubscriptionStateHandlerImpl.java

```diff
@@ -0,0 +1,117 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.mantisrx.server.worker;
+
+import io.mantisrx.server.master.client.MantisMasterGateway;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import rx.Subscription;
+
+class SubscriptionStateHandlerImpl implements SinkSubscriptionStateHandler {
+
+    private static final Logger logger = LoggerFactory.getLogger(SubscriptionStateHandlerImpl.class);
+    private final AtomicReference<ScheduledFuture> timedOutExitFutureRef = new AtomicReference<>();
+    private final String jobId;
+    private final MantisMasterGateway masterClientApi;
+    private ScheduledThreadPoolExecutor executor;
+    private final long subscriptionTimeoutSecs;
+    private final long minRuntimeSecs;
+    private long startedAt = System.currentTimeMillis();
+
+    SubscriptionStateHandlerImpl(String jobId, MantisMasterGateway masterClientApi, long subscriptionTimeoutSecs, long minRuntimeSecs) {
+        this.jobId = jobId;
+        this.masterClientApi = masterClientApi;
+        this.subscriptionTimeoutSecs = subscriptionTimeoutSecs;
+        this.minRuntimeSecs = minRuntimeSecs;
+        executor = this.subscriptionTimeoutSecs > 0L ? new ScheduledThreadPoolExecutor(1) : null;
+    }
+
+    @Override
+    public synchronized void start() {
+        startedAt = System.currentTimeMillis();
+        onSinkUnsubscribed(); // start off as unsubscribed
+    }
+
+    private long evalSubscriberTimeoutSecs() {
+        return Math.max(
+                minRuntimeSecs - ((System.currentTimeMillis() - startedAt) / 1000L),
+                subscriptionTimeoutSecs
+        );
+    }
+
+    @Override
+    public synchronized void onSinkUnsubscribed() {
+        if (executor == null)
+            return;
+        if (timedOutExitFutureRef.get() == null) {
+            timedOutExitFutureRef.set(
+                    executor.schedule(
+                            () -> {
+                                final AtomicReference<Throwable> error = new AtomicReference<Throwable>();
+                                while (true) {
+                                    logger.info("Calling master to kill due to subscription timeout");
+                                    final Subscription subscription = masterClientApi.killJob(jobId, "MantisWorker", "No subscriptions for " +
+                                            subscriptionTimeoutSecs + " secs")
+                                            .subscribe();
+                                    // wait for kill to happen, we won't be running if it succeeds
+                                    try {Thread.sleep(60000);} // arbitrary sleep before we retry the kill
```

Comment:
formatting and also a constant preferably configuration.

Reply:
As before, this is existing code actually. If you refresh the PR, you should be able to see it now.